Resource updates to support WCOSS2

KateFriedman-NOAA opened the pull request:
**Description**

This PR primarily brings back resource updates from testing on WCOSS2. These updates have been tested on Orion and further adjusted as needed. Testing on Hera continues but is ok so far. There are also changes to add `BASE_CPLIC` for WCOSS2.

Changes:

1. pull in some GFSv16.3 adjustments (mainly from NCO reverting memory increases)
2. add WCOSS2 `BASE_CPLIC` to `config.coupled_ic`
3. add `make_nsstbufr` and `make_nsstbufr` to `wcoss2.yaml` hosts file to match S4 updates to other hosts files
4. remove errant `)` in `wcoss2.yaml` hosts file
5. updates to `config.fv3`
* add WCOSS2 `npe_node_max` 
* change `WRTTASK_PER_GROUP` from `$npe_node_max` to 64 for every resolution (half of node size on WCOSS2)
* add check for `WRTTASK_PER_GROUP` to set it back to `$npe_node_max` if 64 is greater than $npe_node_max on machine (Hera/Orion)
* change C192 `nth_fv3` from 2 to 1 (2 fails on WCOSS2 but 1 works everywhere, further examine this on WCOSS2 during optimization work)
* change C384 `WRITE_GROUP` from 1 to 2 (from WCOSS2 GFSv16 port work, works everywhere)
* move C384 from `netcdf_parallel` to `netcdf` (serial) case check at bottom; need to run C384 with serial netcdf on WCOSS2 (needs further investigation)
6. updates to `config.resources`
* add WCOSS2 `npe_node_max`
* add {} around variables
* replace `--exclusive` with `is_exclusive=True`
* update prep job to run exclusively on WCOSS2 but set 40GB memory elsewhere
* pull in WCOSS2 GFSv16 port resource updates:
  * add `_gfs` versions of some job resource variables where needed to define gdas and gfs suite versions of jobs differently
  * adjust some walltimes
  * add `npe_node_cycle` where needed for WCOSS2 `APRUN_CYCLE` launcher commands
  * add memory settings for non-exclusive jobs (setting `is_exclusive=True` otherwise)
  * update some `npe_node_*` values to either be `npe_node_max / thread #` or if greater than `npe_node_max` set it to `npe_node_max` (for R&Ds)
  * reduce `npe_wavepostbndpnt` from 280 to 240
  * increase `npe_wavepostbndpntbll` from 280 to 448
  * reduce `npe_wavepostpnt` from 280 to 200
  * change `npe_wavegempak` from `npe_node_max` to 1
  * change `npe_waveawipsbulls` from `npe_node_max` to 1
  * change `npe_waveawipsgridded` from `npe_node_max` to 1
  * adjust C768 and C384 analysis job resources
  * reduce `npe_analdiag` from 112 to 96 and add note `# Should be at least twice npe_ediag`
  * increase `npe_gldas` from 96 to 112 and change `npe_node_gaussian` to be `npe_node_max / nth_gaussian`
  * increase `npe_post` from 112 to 126 and set `npe_node_post*` values to `npe_post`
  * change `npe_wafsgrib2` from 1 to 18
  * change `npe_wafsgrib20p25` from 1 to 11
  * set `memory_echgres` to 200GB on WCOSS2
  * reduce `npe_ediag` from 56 to 48
  * add WCOSS2 blocks to eupd section, use ops resources for C768 and C384; optimization is needed; also update Hera settings to stop using 40 threads (npe_node_max)
  * reduce `nth_ecen` from 6 to 4 on WCOSS2 and Orion
  * reduce `nth_epos` from 6 to 4 on WCOSS2 and Orion
  * update postsnd resources everywhere
  * reduce awips resources to 1 npe, 1 npe_node, and 1 thread
  * update gempak resources

Resource changes in generated xmls on R&Ds:

* all
  * post job nodes reduced from 10 to 4
  * some ppn values reduced from 40 to a smaller non-npe_node_max value (no node # changes for these jobs though)
  * gdasfcst nodes reduced from 10 to 5
  * gdasepos nodes reduced from 14 to 8 on Orion
* coupled C384
  * gfswavepostpnt nodes reduced from 7 to 4 

Notes: 

1. non-atmos-only and non-coupled jobs were not tested on WCOSS2, downstream ops-only jobs (e.g. gempak/awips) were also not tested for this PR
2. work is still needed for METplus and fit2obs on WCOSS2
3. S2SWA is not supported on WCOSS2 (waiting for mapl)
4. ocnpost does not work on WCOSS2

**Type of change**

Updates for supporting WCOSS2.

**How Has This Been Tested?**

- [x] Clone and Build tests on WCOSS2, Hera, and Orion
- [x] Cycled atmos-only C768C384L127 tests on WCOSS2
- [x] Cycled atmos-only C384C192L127 and C192C96L127 tests on WCOSS2, Hera, Orion
- [x] Forecast-only S2SW on WCOSS2 and Orion

Some continued testing is happening on Hera, Orion, and WCOSS2.
  
Refs #419

KateFriedman-NOAA committed:
Add WCOSS2 BASE_CPLIC to config.coupled_ic

- WCOSS2 BASE_CPLIC is /lfs/h2/emc/global/noscrub/emc.global/IC/COUPLED

Refs #419

KateFriedman-NOAA committed:
Remove errant ) from WCOSS2 hosts file for COMINsyn

Refs #419

KateFriedman-NOAA committed:
Updates to config.fv3 from WCOSS2 testing

- Add npe_node_max=128 for WCOSS2 in machine if-block.
- Change WRTTASK_PER_GROUP and WRTTASK_PER_GROUP_GFS to be 64 for all
supported resolutions (good value for WCOSS2) but then add check for
whether WRTTTASK_PER_GROUP* variables are greater than npe_node_max
value (for R&DS) and if so, set WRTTASK_PER_GROUP* values to equal the
npe_node_max value (prior setting).
- Change C384 DELTIM to 200 (current GFSv16 ops value).

Refs #419

KateFriedman-NOAA committed:
Adjust how S2SW resource defaults are used

- Rename the resource variables in config.defaults.s2sw to include
"_s2sw" in the names.
- Update the C384 block of config.fv3 to use the "_s2sw" variables for
the relevant resource variable values (when set).
- When not running coupled S2Sw the "_s2sw" variables will not be set
and the C384 block of config.fv3 will not use them but instead use the
appropriate values for non-coupled.
- Resolves issue where S2Sw resource values were being forced onto the
C384 atmos-only enkf forecast jobs.

Refs #419

KateFriedman-NOAA committed:
Add "is_exclusive" setting to config.resources

- When job should run exclusively, set is_exclusive=True.
- Replace native_*="--exclusive" with is_exclusive=True.

Refs #419

KateFriedman-NOAA committed:
Add WCOSS2 to machine npe_node_max check in config.resources

- Set npe_node_max=128 when on WCOSS2.

Refs #419

KateFriedman-NOAA committed:
Update wave job resources in config.resources

- Pull in wave job resources from GFSv16 port.
- Adjust walltimes, tasks, add memory, and add "_gfs" versions of
resource variables as needed.

Refs #419

KateFriedman-NOAA committed:
Update awips and gempak resources in config.resources

- Based on GFSv16 port to WCOSS2.
- Add memory settings for both jobs.
- Adjust tasks and threads for both jobs.
- Add "_gfs" variable versions for gempak job.

Refs #419

KateFriedman-NOAA committed:
Remove duplicate wtime_gempak

Refs #419

KateFriedman-NOAA committed:
Update postsnd job resources in config.resources

- Pull in WCOSS2 GFSv16 ops port resource updates into config.resources.

Refs #419

KateFriedman-NOAA committed:
Change epos thread # from 6 to 4

Refs #419

KateFriedman-NOAA committed:
Update esfc resources in config.resources

- Set $npe_node_esfc to $npe_esfc instead of $npe_node_max; then check if
$npe_node_esfc is greater than $npe_node_max and set to $npe_node_max if
so.
- Set $npe_node_cycle based on $npe_node_max divided by $nth_cycle.
- Add memory setting.

Refs #419

KateFriedman-NOAA committed:
Update ecen resources in config.resources

- Change nth_ecen from 6 to 4.
- Add setting for $npe_node_cycle based on $npe_node_max divided by
$nth_cycle.

Refs #419

KateFriedman-NOAA committed:
Update eupd resources in config.resources

- Remove HERA block for C768.
- Add WCOSS2 block for C768; use GFSv16 ops settings.
- Adjust other resolution HERA blocks to remove npe_eupd settings (use
defaults for all machines) and reduce nth_eupd values to more
appropriate values other than 40 (npe_node_max on R&DS).

Refs #419

KateFriedman-NOAA committed:
Update ediag resources in config.resources

- Reduce npe_ediag from 56 to 48.
- Add memory setting.

Refs #419

KateFriedman-NOAA committed:
Reduce eobs walltime to 15 mins in config.resources

Refs #419

KateFriedman-NOAA committed:
Increase memory for archive jobs

Refs #419

KateFriedman-NOAA committed:
Set memory for echgres in config.resources

Refs #419

KateFriedman-NOAA committed:
Update wafs jobs resources in config.resources

- Add memory settings for wafs jobs.
- Adjust task values based on GFSv16 ops port.
- Set $npe_node_wafs* to $npe_wafs* values.

Refs #419

KateFriedman-NOAA committed:
Update post resources in config.resources

- Add "_gfs" variables.
- Reduce walltimes.
- Increase npe_post from 112 to 126 (GFSv16 ops port value).
- Check if $npe_node_post[_gfs] is greater than $npe_node_max and set it
to $npe_node_max if so.

Refs #419

KateFriedman-NOAA committed:
Increase gdasfcst walltime to 40mins in config.resources

Refs #419

KateFriedman-NOAA committed:
Update gldas job resources in config.resources

- Increase $npe_gldas from 96 to 112.
- Set $npe_node_gldas to $npe_gldas instead of $npe_node_max.
- Set $npe_node_gaussian based on $npe_node_max divided by
$nth_gaussian.

Refs #419

KateFriedman-NOAA committed:
Update analdiag resources in config.resources

- Reduce $npe_analdiag from 112 to 96; add note that $npe_analdiag should
be at least twice npe_diag.
- Set $npe_node_analdiag to $npe_analdiag.
- Check if $npe_node_analdiag is greater than $npe_node_max and set to
$npe_node_max if so.
- Add memory setting.

Refs #419

KateFriedman-NOAA committed:
Update gldas to check $npe_node_gldas

- If $npe_node_gldas is greater than $npe_node_max then set it to
$npe_node_max.

Refs #419

KateFriedman-NOAA committed:
Increase efcs walltime

Refs #419

KateFriedman-NOAA committed:
Update anal job resources in config.resources

- Bring in GFSv16 ops port values for C768.
- Also adjust C384 values.
- Add "_gfs" versions of variables so they can be set separately.
- Set $npe_node_cycle based on $npe_node_max divided by $nth_cycle.

Refs #419

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into feature/dev-wcoss2-resources

* upstream/develop:
  Fix companion ocean resolution for C48 (#1066)
  Add trailing slash for gldas topo path (#1064)
  Limit number of CPU for post (#1061)
  Fix eupd trace (#1057)
  Port to S4 (#1023)
  Update to obsproc.v1.0.2 and prepobs.v1.0.1 (#1049)
  Add GDAS to the partial build list (#1050)
  Fix group number being treated as octal in gdas arch (#1053)
  Remove trace from link script (#1046)

KateFriedman-NOAA committed:
Add make_nsstbufr/make_acftbufr to WCOSS2 hosts file

- Set both make_nsstbufr and make_acftbufr to NO.

Refs #419

KateFriedman-NOAA committed:
Resources updates in config.fv3 from WCOSS2 testing

- add if-block for WCOSS2 to set export nth_fv3_efcs=1; nth_fv3_efcs is
new and used for nth_fv3 in the efcs job
- change C384 WRITE_GROUP from 1 to 2
- change C384 to run serially

Refs #419

KateFriedman-NOAA committed:
WCOSS2 updates from testing in config.resources

- set npe_node_analdiag=$npe_analdiag
- add WCOSS2 if-block to set memory_echgres="200GB" only on WCOSS2
- add npe_node_cycle back into ecen block
- update nth_efcs to use nth_fv3_efcs if set; matching update in
config.fv3

Refs #419

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Updates from R&D testing

- remove nth_fv3_efcs and use nth_fv3=1 for C192 everywhere
- set nth_epos=6 on Hera

Refs #419

KateFriedman-NOAA committed:
Update anal block in config.resources

- add WCOSS2 check to set general threads to 8 and keep it 5 threads
on R&Ds to keep nodes the same on R&Ds

Refs #419

KateFriedman-NOAA committed:
Merge branch 'feature/dev-wcoss2-resources' of https://github.com/KateFriedman-NOAA/global-workflow into feature/dev-wcoss2-resources

* 'feature/dev-wcoss2-resources' of https://github.com/KateFriedman-NOAA/global-workflow:
  Updates from R&D testing
  WCOSS2 updates from testing in config.resources
  Resources updates in config.fv3 from WCOSS2 testing
  Modify default file location for monitor data when using rocoto (#1065)

KateFriedman-NOAA committed:
Update C384 nth_fv3 and remove _s2sw

- Update C384 nth_fv3 from 2 to 1.
- Remove added "_s2sw" text to S2SW and C384 resource variables.

Refs #419

KateFriedman-NOAA committed:
Add npe_node_max check for wavepostpnt

Refs #419

WalterKolczynski-NOAA commented on parm/config/config.resources:
```diff
@@ -67,72 +73,88 @@ elif [ ${step} = "waveinit" ]; then
     export npe_waveinit=12
     export nth_waveinit=1
-    export npe_node_waveinit=$(echo "${npe_node_max} / ${nth_waveinit}" | bc)
+    export npe_node_waveinit=$npe_waveinit
```
Won't some of these change if the number of threads is increased later?

KateFriedman-NOAA replied:
Some of these were brought in from the GFSv16 WCOSS2 port. I reviewed the `npe_node_*` variables and made adjustments. These adjustments did not change the xml resources for a high-res cycled test on WCOSS2 so that's good. The wave jobs aren't run in testing for this PR so we'll need to check them when we turn waves back on in the future. Similar note for the wafs jobs.

I also checked that the xml resources don't change on Hera with these updates. They don't.

I also pulled in recent updates from GFSv16.3 (ecf scripts and memory reductions).

Please review the `config.resources` changes.

WalterKolczynski-NOAA reviewed (changes requested):
Confused about these `npe_node_*` changes in general. Values should never be larger than the core size, otherwise you might put multiple ranks on the same CPU.

aerorahul commented on parm/config/config.resources:
```diff
@@ -67,72 +73,88 @@ elif [ ${step} = "waveinit" ]; then
 
-    export wtime_waveawipsgridded="00:30:00"
-    export npe_waveawipsgridded=${npe_node_max}
+    export wtime_waveawipsgridded="02:00:00"
```
that is a huge increase.

KateFriedman-NOAA replied:
This was a change from GFSv16 ops: https://github.com/NOAA-EMC/global-workflow/blob/operations/ecf/scripts/gfs/wave/post/jgfs_wave_prdgen_gridded.ecf#L6

aerorahul commented on parm/config/config.resources:
```diff
@@ -206,13 +229,21 @@ elif [ ${step} = "aeroanlfinal" ]; then
 
-    export wtime_anal="01:00:00"
-    export npe_anal=1000
+    export wtime_anal="00:50:00"
+    export wtime_anal_gfs="00:40:00"
```
I understand the granularity, but 10m is hardly any thing. So would keeping `gfs` the same as `gdas` have meaningful impact?

KateFriedman-NOAA replied:
In development, no. We'll just need to revisit these walltimes for GFSv17 ops (these are the current ops walltimes). Want me to change the `gfs` value to be 50mins?

aerorahul reviewed (commented):
most of the changes look good.
a couple of questions.

KateFriedman-NOAA committed:
Resource updates

- Pull in GFSv16.3 updates from NCO and into ecf scripts.
- Adjust npe_node variables for some jobs to better adjust to different
npe_node_max values.
- Address some linter warnings about {}.

Refs #419

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Increase wafs blending memory to 15GB

Refs #419

aerorahul commented on parm/config/config.fv3:
```diff
@@ -52,9 +54,11 @@ case $case_in in
         export cdmbgwd="0.071,2.1,1.0,1.0"  # mountain blocking, ogwd, cgwd, cgwd src scaling
         export WRITE_GROUP=1
-        export WRTTASK_PER_GROUP=$npe_node_max
+        export WRTTASK_PER_GROUP=64
```
Can you explain why 64 is hard-wired here?  Isn't `npe_node_max=40` on Orion, in which case this is going to be an issue. No?
I think you are trying to set this so that on WCOSS it won't be requesting 128 cores. Correct?

KateFriedman-NOAA replied:
During the port of GFSv16 to WCOSS2 George V and I found that WRTTASK_PER_GROUP=64 was good for the resolutions I tested. I set `WRTTASK_PER_GROUP=64` for all resolutions as the default but also include a check in the lines after each of these to set it back to `$npe_node_max` (what it was before) when `WRTTASK_PER_GROUP > npe_node_max` (on the other platforms). Example:
```
if [[ "$WRTTASK_PER_GROUP" -gt "$npe_node_max" ]]; then export WRTTASK_PER_GROUP=$npe_node_max ; fi
```
So the values for the R&Ds is preserved and WCOSS2 is considered. The same thing was done in the `dev_v16.2.x` branch:
https://github.com/NOAA-EMC/global-workflow/blob/dev_v16.2.x/parm/config/config.fv3.emc.dyn

aerorahul commented on parm/config/config.fv3:
```diff
@@ -79,28 +85,32 @@ case $case_in in
         export cdmbgwd="0.23,1.5,1.0,1.0"  # mountain blocking, ogwd, cgwd, cgwd src scaling
         export WRITE_GROUP=1
-        export WRTTASK_PER_GROUP=$npe_node_max
+        export WRTTASK_PER_GROUP=64
```
I think I must be confused about this.

KateFriedman-NOAA replied:
See comment reply above.

aerorahul reviewed (commented):
I think I need more explanations on the hardwiring of 64 across for all platforms.

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Increase esfc memory back up to 80GB

- Received an OOM kill when testing 60GB. Increased back up to 80GB and the job completed.
- This is the only job still encountering a "Cgroup mem" warning however. Further investigation is needed.

Refs #419

KateFriedman-NOAA committed:
Address SC2250 linter warnings

Refs #419

KateFriedman-NOAA committed:
Add partition_service to WCOSS2 hosts file

Matches updates made to other hosts files recently.

Refs #419

KateFriedman-NOAA committed:
Increase esfc memory to 80GB

Refs #419

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources

KateFriedman-NOAA committed:
Update build_ufs.sh to force S2SW on WCOSS2

- Aerosols not yet supported on WCOSS2 so can't build with S2SWA app there.
- Force UFS to build with the S2SW app on WCOSS2 for now.

Refs #419

KateFriedman-NOAA committed:
Change WCOSS2 to lowercase wcoss2 in build_ufs.sh

Refs #419

KateFriedman-NOAA committed:
Set hpssarch to NO by default on WCOSS2

- Update WCOSS2 hosts yaml file to set `hpssarch` to "NO" by default.
- Limited bandwidth between WCOSS2 to HPSS so don't want users archiving to HPSS unless necessary.

Refs #419

KateFriedman-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature/dev-wcoss2-resources
